python/ray/worker.py

```diff
@@ -589,93 +575,22 @@ def submit_task(self,
-            if execution_dependencies is None:
-                execution_dependencies = []
-
-            if new_actor_handles is None:
-                new_actor_handles = []
-
-            if driver_id is None:
-                driver_id = self.task_driver_id
-
-            if resources is None:
-                raise ValueError("The resources dictionary is required.")
-            for value in resources.values():
-                assert (isinstance(value, int) or isinstance(value, float))
-                if value < 0:
-                    raise ValueError(
-                        "Resource quantities must be nonnegative.")
-                if (value >= 1 and isinstance(value, float)
-                        and not value.is_integer()):
-                    raise ValueError(
-                        "Resource quantities must all be whole numbers.")
-
-            if placement_resources is None:
-                placement_resources = {}
-
-            # Increment the worker's task index to track how many tasks
-            # have been submitted by the current task so far.
-            self.task_context.task_index += 1
-            # The parent task must be set for the submitted task.
-            assert not self.current_task_id.is_nil()
-            # Current driver id must not be nil when submitting a task.
-            # Because every task must belong to a driver.
-            assert not self.task_driver_id.is_nil()
-            # Submit the task to raylet.
-            function_descriptor_list = (
-                function_descriptor.get_function_descriptor_list())
-            assert isinstance(driver_id, DriverID)
-            task = ray._raylet.Task(
-                driver_id,
-                function_descriptor_list,
-                args_for_raylet,
-                num_return_vals,
-                self.current_task_id,
-                self.task_context.task_index,
-                actor_creation_id,
-                actor_creation_dummy_object_id,
-                max_actor_reconstructions,
-                actor_id,
-                actor_handle_id,
-                actor_counter,
-                new_actor_handles,
-                execution_dependencies,
-                resources,
-                placement_resources,
-            )
-            self.raylet_client.submit_task(task)
-
-            return task.returns()
+        self.task_context.task_index += 1
+        kwargs["driver_id"] = kwargs.get("driver_id", self.task_driver_id)
+        if kwargs["profile"]:
+            with profiling.profile("submit_task"):
+                return self.raylet_client.submit_task(
+                    function_descriptor_list, args, self.current_task_id,
+                    self.task_context.task_index, put, **kwargs)
+        else:
+            return self.raylet_client.submit_task(
+                function_descriptor_list, args, self.current_task_id,
+                self.task_context.task_index, put, **kwargs)
```

Comment:
we can avoid code duplication in this way:
```
with profiling.profile("submit_task", enable=kwargs["profile"]):
   ...
```
when `enable=False`, the profile method returns a `NULL_LOG_SPAN`

Also, do we want enable profiling per submit? I think it makes more sense to have a global flag.

Reply:
Thanks, I fixed both!

Reply:
actually, calling the contextmanager introduces too much overhead, so I removed that again.